java/src/main/java/io/cucumber/gherkin/PickleCompiler.java

```diff
@@ -98,25 +98,34 @@ private void compileRule(List<Pickle> pickles, Rule rule, List<Tag> parentTags,
     }
 
     private void compileScenario(List<Pickle> pickles, Scenario scenario, List<Tag> parentTags, List<Step> backgroundSteps, String language, String uri) {
-        List<PickleStep> steps = new ArrayList<>();
+        int pickleStepCount = scenario.getSteps().size() + backgroundSteps.size();
+        List<PickleStep> steps = new ArrayList<>(pickleStepCount);
         if (!scenario.getSteps().isEmpty()) {
-            List<Step> allSteps = new ArrayList<>();
+            // usually, a scenario has at least one step, but that's not mandatory
+            List<Step> allSteps = new ArrayList<>(pickleStepCount);
             allSteps.addAll(backgroundSteps);
             allSteps.addAll(scenario.getSteps());
 
             StepKeywordType lastKeywordType = StepKeywordType.UNKNOWN;
             for (Step step : allSteps) {
                 StepKeywordType stepKeywordType = step.getKeywordType().get();
-                if (stepKeywordType != StepKeywordType.CONJUNCTION)
+                if (stepKeywordType != StepKeywordType.CONJUNCTION) {
                     lastKeywordType = stepKeywordType;
+                }
 
                 steps.add(pickleStep(step, lastKeywordType));
             }
         }
 
-        List<Tag> scenarioTags = new ArrayList<>();
-        scenarioTags.addAll(parentTags);
-        scenarioTags.addAll(scenario.getTags());
+        List<Tag> scenarioTags;
+        if (parentTags.isEmpty() && scenario.getTags().isEmpty()) {
+            // in most cases, the scenario will have no tag, so we avoid creating a new list
+            scenarioTags = emptyList();
```

Comment:
You could  `scenarioTags = parentTags` and remove the `parentTags.isEmpty()` check.